command/build.go

```diff
@@ -88,36 +89,19 @@ func (c *BuildCommand) RunContext(buildCtx context.Context, cla *BuildArgs) int
 		return ret
 	}
 
-	diags = TrySetupHCP(packerStarter)
-	ret = writeDiags(c.Ui, nil, diags)
-	if ret != 0 {
-		return ret
-	}
-
-	// This build currently enforces a 1:1 mapping that one publisher can be assigned to a single packer config file.
-	// It also requires that each config type implements this ConfiguredArtifactMetadataPublisher to return a configured bucket.
-	// TODO find an option that is not managed by a globally shared Publisher.
-	ArtifactMetadataPublisher, diags := packerStarter.ConfiguredArtifactMetadataPublisher()
-	if diags.HasErrors() {
-		return writeDiags(c.Ui, nil, diags)
+	hcpHandler, err := hcp.GetHCPHandler(packerStarter)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("HCP setup failed: %s", err))
+		return -1
 	}
 
-	// We need to create a bucket and an empty iteration before we retrieve builds
-	// so that we can add the iteration ID to the build's eval context
-	if ArtifactMetadataPublisher != nil {
-		if err := ArtifactMetadataPublisher.Initialize(buildCtx); err != nil {
-			diags := hcl.Diagnostics{
-				&hcl.Diagnostic{
-					Summary:  "HCP Packer Registry iteration initialization failed",
-					Detail:   fmt.Sprintf("Failed to initialize iteration for %q\n %s", ArtifactMetadataPublisher.Slug, err),
-					Severity: hcl.DiagError,
-				},
-			}
-			return writeDiags(c.Ui, nil, diags)
-		}
+	err = hcpHandler.PopulateIteration(buildCtx)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("HCP populating iteration failed: %s", err))
+		return -1
 	}
 
-	builds, diags := packerStarter.GetBuilds(packer.GetBuildsOptions{
+	builds, hcpMap, diags := packerStarter.GetBuilds(packer.GetBuildsOptions{
```

Comment:
If we get the list of build names here. What do we populate the iteration with? 

Is it an empty iteration with no defined builds?

Reply:
The map is here for HCL templates with a `name` set for the `build` block.
If this is set, the builds are populated on HCP using the source name, which is not exposed in the `packersdk.Build` object returned by `GetBuilds`, hence that function that maps the returned build's name (which includes the `name` from the block), to the HCP-created build.
It's hack-ish, but at the moment I'm not sure what would be a cleaner way to handle this without causing any kind of surprises to the users. If we start using the generated name everywhere, this'll change the names of the builds on HCP, which I'm not sure we want to do (although this would be more accurate).